File: python/kakeibo/allocate_amounts.py
"""Allocate receipt amount across categories (mirror of allocateAmounts.ts)."""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from .types import (
    AmountAllocationResult,
    Category,
    CategoryAllocation,
    ReceiptBreakdown,
)


@dataclass(frozen=True)
class AllocateOptions:
    per_item_amounts: Optional[list[float]] = None


def allocate_amounts_by_category(
    breakdown: ReceiptBreakdown,
    total_amount: float,
    options: Optional[AllocateOptions] = None,
) -> AmountAllocationResult:
    opts = options or AllocateOptions()
    _assert_non_negative(total_amount, "total_amount")

    items = breakdown.items
    if len(items) == 0:
        return AmountAllocationResult(
            total_amount=total_amount,
            allocations=[],
            strategy="equal_split",
            unclassified_amount=total_amount,
        )

    per_item = _resolve_per_item_amounts(len(items), total_amount, opts.per_item_amounts)
    strategy: str = "per_item_amounts" if opts.per_item_amounts is not None else "equal_split"

    buckets: dict[Optional[Category], dict] = {}
    for index, item in enumerate(items):
        bucket = buckets.setdefault(item.category, {"amount": 0.0, "item_texts": []})
        bucket["amount"] += per_item[index]
        bucket["item_texts"].append(item.text)

    allocations = [
        CategoryAllocation(category=cat, amount=_round_currency(b["amount"]), item_texts=list(b["item_texts"]))
        for cat, b in buckets.items()
    ]
    allocations.sort(key=lambda a: a.amount, reverse=True)

    unclassified_amount = next(
        (a.amount for a in allocations if a.category is None), 0.0
    )

    return AmountAllocationResult(
        total_amount=total_amount,
        allocations=allocations,
        strategy=strategy,  # type: ignore[arg-type]
        unclassified_amount=unclassified_amount,
    )


def _resolve_per_item_amounts(
    item_count: int,
    total_amount: float,
    provided: Optional[list[float]],
) -> list[float]:
    if provided is not None:
        if len(provided) != item_count:
            raise ValueError(
                f"per_item_amounts length ({len(provided)}) "
                f"does not match items length ({item_count})"
            )
        for amount in provided:
            _assert_non_negative(amount, "per_item_amounts entry")
        return list(provided)
    return _equal_split(total_amount, item_count)


def _equal_split(total: float, count: int) -> list[float]:
    if count == 0:
        return []
    base = math.floor((total / count) * 100) / 100
    amounts = [base] * count
    distributed = _round_currency(base * count)
    remainder = _round_currency(total - distributed)
    amounts[-1] = _round_currency(amounts[-1] + remainder)
    return amounts


def _round_currency(value: float) -> float:
    return round(value * 100) / 100


def _assert_non_negative(value: float, label: str) -> None:
    if value is None or not math.isfinite(value) or value < 0:
        raise ValueError(f"{label} must be a finite non-negative number, got {value}")

File: src/allocateAmounts.ts
import type {
  AmountAllocationResult,
  Category,
  CategoryAllocation,
  ReceiptBreakdown
} from "./types";

export type AllocateOptions = {
  perItemAmounts?: number[];
};

export function allocateAmountsByCategory(
  breakdown: ReceiptBreakdown,
  totalAmount: number,
  options: AllocateOptions = {}
): AmountAllocationResult {
  assertNonNegative(totalAmount, "totalAmount");

  const items = breakdown.items;
  if (items.length === 0) {
    return {
      totalAmount,
      allocations: [],
      strategy: "equal_split",
      unclassifiedAmount: totalAmount
    };
  }

  const perItem = resolvePerItemAmounts(items.length, totalAmount, options.perItemAmounts);
  const strategy = options.perItemAmounts ? "per_item_amounts" : "equal_split";

  const buckets = new Map<Category | null, { amount: number; itemTexts: string[] }>();
  for (let i = 0; i < items.length; i++) {
    const item = items[i];
    const amount = perItem[i];
    const bucket = buckets.get(item.category) ?? { amount: 0, itemTexts: [] };
    bucket.amount += amount;
    bucket.itemTexts.push(item.text);
    buckets.set(item.category, bucket);
  }

  const allocations: CategoryAllocation[] = Array.from(buckets.entries())
    .map(([category, bucket]) => ({
      category,
      amount: bucket.amount,
      itemTexts: bucket.itemTexts
    }))
    .sort((a, b) => b.amount - a.amount);

  const unclassifiedAmount = allocations.find((a) => a.category === null)?.amount ?? 0;

  return {
    totalAmount,
    allocations,
    strategy,
    unclassifiedAmount
  };
}

function resolvePerItemAmounts(
  itemCount: number,
  totalAmount: number,
  provided: number[] | undefined
): number[] {
  if (provided) {
    if (provided.length !== itemCount) {
      throw new Error(
        `perItemAmounts length (${provided.length}) does not match items length (${itemCount})`
      );
    }
    for (const amount of provided) {
      assertNonNegative(amount, "perItemAmounts entry");
    }
    return provided.slice();
  }

  return equalSplit(totalAmount, itemCount);
}

function equalSplit(total: number, count: number): number[] {
  if (count === 0) return [];
  const base = Math.floor((total / count) * 100) / 100;
  const amounts = new Array<number>(count).fill(base);
  const distributed = roundCurrency(base * count);
  const remainder = roundCurrency(total - distributed);
  amounts[count - 1] = roundCurrency(amounts[count - 1] + remainder);
  return amounts;
}

function roundCurrency(value: number): number {
  return Math.round(value * 100) / 100;
}

function assertNonNegative(value: number, label: string): void {
  if (!Number.isFinite(value) || value < 0) {
    throw new Error(`${label} must be a finite non-negative number, got ${value}`);
  }
}
